src/olympia/addons/views.py

```diff
@@ -737,3 +740,68 @@ def get_queryset(self):
     def as_view(cls, **kwargs):
         view = super(AddonSearchView, cls).as_view(**kwargs)
         return non_atomic_requests(view)
+
+
+class AddonFeaturedView(GenericAPIView):
+    authentication_classes = []
+    permission_classes = []
+    serializer_class = AddonSerializer
+    # We accept the 'page_size' parameter but we do not allow pagination for
+    # this endpoint since the order is random.
+    pagination_class = None
+    queryset = Addon.objects.valid()
+
+    def get(self, request, *args, **kwargs):
+        queryset = self.filter_queryset(self.get_queryset())
+        serializer = self.get_serializer(queryset, many=True)
+
+        # Simulate pagination-like results, without actual pagination.
+        return Response({'results': serializer.data})
+
+    @classmethod
+    def as_view(cls, **kwargs):
+        view = super(AddonFeaturedView, cls).as_view(**kwargs)
+        return non_atomic_requests(view)
+
+    def filter_queryset(self, queryset):
+        # We can pass the optional lang parameter to either get_creatured_ids()
+        # or get_featured_ids() below to get locale-specific results in
+        # addition to the generic ones.
+        lang = self.request.GET.get('lang')
+        if 'category' in self.request.GET:
+            # If a category is passed then the app and type parameters are
+            # mandatory because we need to find a category in the constants to
+            # pass to get_creatured_ids(), and category slugs are not unique.
+            # AddonCategoryFilterParam parses the request parameters for us to
+            # determine the category.
+            try:
+                category = AddonCategoryFilterParam(self.request).get_value()
+            except ValueError:
+                raise ParseError(
+                    'Invalid app, category and/or type parameter(s).')
+            ids = get_creatured_ids(category, lang)
+        else:
+            # If no category is passed, only the app parameter is mandatory,
+            # because get_featured_ids() needs it to find the right collection
+            # to pick addons from. It can optionally filter by type, so we
+            # parse request for that as well.
+            try:
+                app = AddonAppFilterParam(
+                    self.request).get_object_from_reverse_dict()
+                type_ = None
```

Comment:
Yeah, it's a bit unfortunate. The reason I have to do this is because `type` is optional and `AddonTypeFilterParam` currently raises `ValueError` if you call `get_value()` and it's invalid or absent. I do want the exception if it's invalid, but in this particular case and only in this particular case I don't want the exception if it's absent.

I guess I'll need to refactor `AddonTypeFilterParam` and its siblings to accept some kind of `optional` argument to control this behavior. Then I could remove the boilerplate.